### PR TITLE
New client - possible first release

### PR DESCRIPTION
**Pull request by stephens-reservoir**
This is a pretty significant amount of refactoring and minor improvements for the new client. I have removed the old code, except for a couple files in legacy/ that are still useful to keep around, as they haven't necessarily been converted, although they aren't used currently. I've split into multiple files, created tests and documentation, and overall left it in a state where it's fine for me to move on to something else if needed.

### CODE REVIEW

File: tests/configs/matmult.meta [hunk 1]
```diff
@@ -0,0 +1,14 @@
+MORPH_SELECT_SMP_MATMULT {
```
Comment (meister-reservoir) [inline tests/configs/matmult.meta:1]: A note on the motivation for JSON: I think the .meta version is actually more compact and easier to read than the JSON one. The JSON one is still readable, though. But I think the true advantage perhaps is being able to serialize and deserialize it more easily.

File: tests/test_main.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import os
```
Comment (meister-reservoir) [inline tests/test_main.py:1]: Could you kindly port these tests to R-QA please?

File: tests/test_main.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    def temp_dir():
+        return arcc_tmp_dir
+    generated_file = src.production.rstream_production("", build_cmd, temp_dir)
+    # make sure it exists, and has contents
+    assert os.path.exists(generated_file)
+    assert os.path.getsize(generated_file) > 0
+    # make sure it parses
+    import src.meta_data_parser
+    src.meta_data_parser.parse(generated_file)
+    # NOTE: ideally, this would assert some properties on the generated file
```
Comment (meister-reservoir) [inline tests/test_main.py:28]: That's totally fine I think.

File: tests/configs/matmult.json [hunk 4]
```diff
@@ -0,0 +1,94 @@
+{
+  "global_env": {
```
Comment (meister-reservoir) [inline tests/configs/matmult.json:2]: I humbly suggest you beef up the format documentation with a few simple examples.

File: README.md [hunk 5]
```diff
@@ -1,3 +1,54 @@
...
+
+To run the tests, run `tests/test`. If there is no output, then all the tests passed. 
+
+Use the `--verbose` flag for either command for more detailed run information.
+
+# Format
+
+There are two formats - one of them is a legacy format used currently by R-Stream, the other is a newer format that's preferred which will be described below. The arguments to be tuned are leaves of a tree, and the elements of that tree can be used to help organize the formatting and constraints for those arguments. See `tests/configs` for some examples. 
+
+The high-level format is:
```
[the remark this answers is not in the document]
Comment (stephens-reservoir) [reply README.md:19]: Ah, sorry. I don't mean that the format is high-level, it's just my weird way of saying "the format is...". Fixed

File: com/reservoir/arcc/ARCC.java [hunk 6]
```diff
@@ -1,337 +0,0 @@
-// DOE-GPU rights
```
[the remark this answers is not in the document]
Comment (stephens-reservoir) [reply com/reservoir/arcc/ARCC.java:1]: I'm not sure what you mean. It doesn't need it

File: tests/test_main.py (hunk 2 above)
Comment (stephens-reservoir) [reply tests/test_main.py:1]: Added jira task for it

File: README.md [hunk 7]
```diff
@@ -1,3 +1,54 @@
...
+
+# Running + Testing
+
+Run `bin/arcc --help` for information about how to run arcc. For ane example, run `bin/arcc --config tests/configs/sleep.json` to show arcc "tuning" a dummy compiler that simply generates a script that runs for a tunable amount of time. The `bin` dir can be safely added to your path to make running `arcc` from anywhere easier. 
+
+To run the tests, run `tests/test`. If there is no output, then all the tests passed. 
+
+Use the `--verbose` flag for either command for more detailed run information.
+
+# Format
```
[the remark this answers is not in the document]
Comment (stephens-reservoir) [reply README.md:15]: Thanks, will add that in.

File: com/reservoir/arcc/ARCC.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (meister-reservoir) [reply com/reservoir/arcc/ARCC.java:1]: I see a few files in R-Stream that import com.reservoir.arcc.ARCC. 
Run a grep -r on 'com.reservoir.arcc.ARCC' to convince yourself.

**Pull request merged**